File: src/compiler/Lexer.lua
--- @class Lexer
--- @field _compiler Compiler
--- @field _kind TokenKind
--- @field _value TokenValue
local Lexer = {}
Lexer.__index = Lexer

--- @alias TokenKind EofTokenKind | LiteralTokenKind | NameTokenKind | TypeNameTokenKind | KeywordTokenKind | PunctuationTokenKind
--- @alias TokenValue nil | boolean | number | string
--- @alias EofTokenKind nil
--- @alias LiteralTokenKind "nil" | "true" | "false" | "<real>" | "<string>"
--- @alias NameTokenKind "<identifier>" | "<intrinsic>"
--- @alias TypeNameTokenKind "Nil" | "Bool" | "Real" | "String"
--- @alias KeywordTokenKind "def" | "if" | "elseif" | "else" | "while" | "break" | "continue" | "pass" | "return"
--- @alias PunctuationTokenKind "(" | ")" | "{" | "}" | "," | ":" | "->" | "=" | "!" | "+" | "-" | "*" | "/" | "%" | "<" | ">" | "==" | "!=" | "<=" | ">=" | "||" | "&&"

Lexer.keywordSet = {
	["nil"] = true,
	["true"] = true,
	["false"] = true,
	["Nil"] = true,
	["Bool"] = true,
	["Real"] = true,
	["String"] = true,
	["def"] = true,
	["if"] = true,
	["elseif"] = true,
	["else"] = true,
	["while"] = true,
	["break"] = true,
	["continue"] = true,
	["pass"] = true,
	["return"] = true,
}

Lexer.punctuationSet = {
	["("] = true,
	[")"] = true,
	["{"] = true,
	["}"] = true,
	[","] = true,
	[":"] = true,
	["="] = true,
	["!"] = true,
	["+"] = true,
	["-"] = true,
	["*"] = true,
	["/"] = true,
	["%"] = true,
	["<"] = true,
	[">"] = true,
	["=="] = true,
	["!="] = true,
	["<="] = true,
	[">="] = true,
	["||"] = true,
	["&&"] = true,
	["->"] = true,
}

--- @param compiler Compiler
--- @return Lexer lexer
function Lexer.new(compiler)
	return setmetatable({
		_compiler = compiler,
		_kind = nil,
		_value = nil,
	}, Lexer)
end

function Lexer:getCompiler()
	return self._compiler
end

--- @return TokenKind kind
--- @return TokenValue value
function Lexer:getToken()
	if self._kind then
		return self._kind, self._value
	else
		return self:nextToken()
	end
end

--- @return TokenKind kind
--- @return TokenValue value
function Lexer:nextToken()
	local reader = self._compiler:getReader()
	local chr = reader:getChar()
	self._kind = nil -- eof
	self._value = nil
	if chr == "#" then -- comment
		-- skip line
		reader:consumeChars("[^\n]")
		reader:nextChar()
		return self:nextToken()
	elseif chr:find("%s") then -- whitespace
		reader:consumeChars("%s")
		return self:nextToken()
	elseif chr:find("[%a_$]") then -- names and keywords
		self._compiler:pushContext()
		self:_scanNamed()
		self._compiler:popContext()
	elseif chr:find("[%d%.]") then -- real literal
		self._compiler:pushContext()
		self:_scanRealLiteral()
		self._compiler:popContext()
	elseif chr == '"' or chr == "'" then -- string literal
		self._compiler:pushContext()
		self:_scanStringLiteral()
		self._compiler:popContext()
	elseif chr:find("%p") then -- punctuation
		self._compiler:pushContext()
		self:_scanPunctuation()
		self._compiler:popContext()
	end
	return self._kind, self._value
end

--- Precondition: current character is a letter, underscore, or intrinsic specifier ("$").
function Lexer:_scanNamed()
	local reader = self._compiler:getReader()
	local intrinsic = reader:getChar() == "$"

	if intrinsic then
		-- skip intrinsic specifier
		reader:nextChar()
	end

	local lexeme = reader:consumeAndQuery("[%w_]")
	local chr = reader:getChar()

	if intrinsic then -- intrinsic
		self._kind = "<intrinsic>"
		self._value = lexeme
	elseif Lexer.keywordSet[lexeme] then -- keyword
		self._kind = lexeme
	else -- identifier
		self._kind = "<identifier>"
		self._value = lexeme
	end

	if chr == "$" then
		self._compiler:addDiagnostic(
			"warning",
			("consider adding a space after %q"):format(intrinsic and "$" .. lexeme or lexeme)
		)
	end
end

--- Precondition: current character is a digit or decimal point.
function Lexer:_scanRealLiteral()
	local reader = self._compiler:getReader()
	local lexeme = reader:consumeAndQuery("[%d%.]")
	local value = tonumber(lexeme)
	self._kind = "<real>"
	self._value = value

	-- ensure the real literal is valid
	if not value then
		self._compiler:addDiagnostic("error", ("malformed real literal %q"):format(lexeme))
		self._value = 0
	end

	-- warn for "weird" token sequence (e.g., 2x, 2.x, 2$)
	local chr = reader:getChar()
	if chr ~= "" and chr:find("[%w$]") then
		self._compiler:addDiagnostic("warning", ("consider adding a space after %q"):format(lexeme))
	end
end

--- Precondition: current character is a quote.
function Lexer:_scanStringLiteral()
	local reader = self._compiler:getReader()
	local quote = reader:getChar()

	-- consume quote
	reader:nextChar()

	-- consume non-quote
	local lexeme = reader:consumeAndQuery("[^" .. quote .. "]")
	self._kind = "<string>"
	self._value = lexeme

	-- expect quote
	if reader:getChar() == quote then
		reader:nextChar()
	else
		self._compiler:addDiagnostic("error", "unterminated string")
	end
end

--- Precondition: current character is punctuation.
function Lexer:_scanPunctuation()
	local reader = self._compiler:getReader()
	local chr = reader:getChar()

	-- greedily match operators
	local op2 = chr .. reader:nextChar()
	if #op2 == 2 and Lexer.punctuationSet[op2] then
		-- match operators of length 2 first
		reader:nextChar()
		self._kind = op2
	elseif Lexer.punctuationSet[chr] then
		-- match operators of length 1 next
		self._kind = chr
	else
		-- unexpected character, recover by scanning the next token
		self._compiler:addDiagnostic("error", ("unexpected character %q"):format(chr))
		return self:nextToken()
	end
end

return Lexer

File: mattylang/lexer.py
from string import ascii_letters, digits, punctuation, whitespace
from typing import Callable, Optional

from mattylang.module import Module


class Token:
    def __init__(self, kind: str, lexeme: str, position: int) -> None:
        self.kind = kind
        self.lexeme = lexeme
        self.position = position

    def __str__(self) -> str:
        if self.kind == 'identifier':
            return f'identifier({self.lexeme})'
        elif self.kind == 'real_literal':
            return f'real_literal({self.lexeme})'
        elif self.kind == 'string_literal':
            return f'string_literal({repr(self.lexeme)})'
        else:
            return self.kind


class Lexer:
    def __init__(self, module: Module):
        self.__module = module
        self.__source = module.source
        self.__position = 0
        self.__token: Optional[Token] = None

    def get_module(self) -> Module:
        return self.__module

    def peek(self):
        if self.__token is None:
            return self.scan()
        return self.__token

    def scan(self):
        self.__token = self.__scan_impl()
        # self.__module.diagnostics.emit_diagnostic('info', f'syntax: scanned {self.__token}', self.__token.position)
        return self.__token

    __space_set = set(whitespace)
    __digit_set = set(digits)
    __id_init_set = set(ascii_letters + '$_')
    __string_char_set = set(ascii_letters + digits + punctuation + ' \t')
    __keyword_set = {'def', 'nil', 'true', 'false', 'if', 'else', 'while',
                     'break', 'continue', 'return', 'Nil', 'Bool', 'Real', 'String'}
    __punctuation = {'(', ')', '{', '}', '=', '!', '+', '-', '*', '/',
                     '%', '<', '>', '==', '!=', '<=', '>=', '||', '&&', '->'}

    def __scan_impl(self) -> Token:
        chr = self.__get_char()
        diagnostics = self.__module.diagnostics
        position = self.__position

        if chr == '':  # eof
            return Token('eof', '', position)
        elif chr in self.__space_set:  # whitespace
            # skip whitespace and continue
            self.__next_char_while(lambda chr: chr in self.__space_set)
            return self.__scan_impl()
        elif chr == '#':  # comment
            # skip comment and continue
            self.__next_char_while(lambda chr: chr != '\n')
            self.__next_char()
            return self.__scan_impl()
        elif chr in self.__id_init_set:  # keyword/identifier
            lexeme = self.__collect_lexeme(lambda chr: chr in self.__id_init_set or chr in self.__digit_set)
            kind = lexeme if lexeme in self.__keyword_set else 'identifier'
            return Token(kind, lexeme, position)
        elif chr in self.__digit_set or chr == '.':  # real literal
            lexeme = self.__collect_lexeme(lambda chr: chr in self.__digit_set)
            kind = 'real_literal'

            if self.__get_char() == '.':
                self.__next_char()
                lexeme = lexeme + '.' + self.__collect_lexeme(lambda chr: chr in self.__digit_set)

            if lexeme == '.':
                diagnostics.emit_diagnostic('error', f'syntax: unexpected character .', position)
                return self.__scan_impl()

            if self.__get_char() in self.__id_init_set or self.__get_char() == '.':
                diagnostics.emit_diagnostic(
                    'warning', f'syntax: expected whitespace after real literal {lexeme}', position)

            return Token('real_literal', lexeme, position)
        elif chr == '"' or chr == "'":  # string literal
            quote = chr
            self.__next_char()
            text = self.__collect_lexeme(lambda chr: chr != quote and chr in self.__string_char_set)

            if self.__get_char() != quote:
                diagnostics.emit_diagnostic(
                    'error', f'syntax: expected {quote} to terminate string', self.__position)
            else:
                self.__next_char()

            return Token('string_literal', text, position)
        else:  # punctuation/other
            # greedily matches longest valid punctuation
            op1 = chr
            op2 = op1 + self.__next_char()
            if len(op2) == 2 and op2 in self.__punctuation:
                self.__next_char()
                return Token(op2, op2, position)
            elif op1 in self.__punctuation:
                return Token(op1, op1, position)
            else:
                # continue if unexpected character
                diagnostics.emit_diagnostic('error', f'syntax: unexpected character {repr(op1)}', position)
                return self.__scan_impl()

    def __get_char(self):
        return self.__source[self.__position] if self.__position < len(self.__source) else ''

    def __next_char(self):
        if self.__position < len(self.__source):
            self.__position += 1
        return self.__get_char()

    def __next_char_while(self, predicate: Callable[[str], bool]):
        while self.__position < len(self.__source) and predicate(self.__get_char()):
            self.__next_char()
        return chr

    def __collect_lexeme(self, predicate: Callable[[str], bool]):
        start = self.__position
        self.__next_char_while(predicate)
        return self.__source[start:self.__position]
